### PR TITLE
feat: add flake with mycelium package

### PR DESCRIPTION
**Pull request by steveej**
* add a flake.nix to package the mycelium binary. with a flake-compat layer consisting of {default,shell}.nix
* add a rudimentary job to the CI workflow that builds the package

### CODE REVIEW

File: flake.nix [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+    nixpkgs,
+    crane,
+    flake-utils,
+  }:
+    flake-utils.lib.eachSystem
+    [
+      flake-utils.lib.system.x86_64-linux
+      flake-utils.lib.system.aarch64-linux
+      flake-utils.lib.system.aarch64-darwin
+    ] (system: let
```
Comment (LeeSmet) [inline flake.nix:22]: Shouldn't this also include x86_64-darwin?

**Comment by LeeSmet** [comment]
I have no experience with nix flakes, so I'll assume this is good. I just have one question after skimming through this.

File: flake.nix (hunk 1 above)
Comment (steveej) [reply flake.nix:22]: thinking about this, i'm not confident about `aarch64-darwin` either as i've not tested it. do you run any CI for darwin?

**Comment by steveej** [comment]
> I have no experience with nix flakes, so I'll assume this is good. I just have one question after skimming through this.

WDYT about building the package exposed by the flake on CI? it could also run the tests. if they don't require WAN access they will most likely just work. if they do we'd have to change a few things so they can reach WAN.

File: flake.nix (hunk 1 above)
Comment (LeeSmet) [reply flake.nix:22]: yes it's included in CI, should be x86_64 architecture though I'm not 100% sure of that

**Comment by LeeSmet** [comment]
I'm all for adding it to CI, right now tests don't really use networking (some tests create peers but those use an in memory connection). Some more in depth tests will probably be added later which do create sockets and connections, but even then I'd say those will only be local to eachother.

To make sure we are on the same page, you mean adding an entirely new CI workflow which builds the package and then  runs the tests, next to the current one which just runs the tests, right?

**Comment by flokli** [comment]
Can you include https://github.com/nix-community/flake-compat, so this is accessible for people who use Nix, but who don't have flakes enabled? It's still an experimental feature, after all ;-)

**Comment by steveej** [comment]
> Can you include https://github.com/nix-community/flake-compat, so this is accessible for people who use Nix, but who don't have flakes enabled? It's still an experimental feature, after all ;-)

added! it's not covered by CI as is, so i'd appreciate if i could rely on you as a reviewer :slightly_smiling_face:

**Comment by LeeSmet** [comment]
Seems there is a linker error on darwin for `ring`: `note: ld: library not found for -liconv`. Possible missing dependancy?

File: flake.nix [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+          cargoExtraArgs = "--bin mycelium";
+
+          nativeBuildInputs = [
+            # required by openssl-sys
+            pkgs.perl
+          ];
+
+          buildInputs = lib.optionals stdenv.isDarwin [
+            darwin.apple_sdk.frameworks.Security
+            darwin.apple_sdk.frameworks.SystemConfiguration
```
Comment (flokli) [inline flake.nix:83]:
```suggestion
            darwin.apple_sdk.frameworks.SystemConfiguration
            pkgs.libiconv
```

**Review by LeeSmet — APPROVED** [review]
Now that CI passes this is good on my end. I presume there are no objections to merge/rebasing this at this point?

**Comment by flokli** [comment]
I was able to invoke `nix-build -A outputs.packages.x86_64-linux.mycelium` and `nix-build -A outputs.packages.aarch64-linux.mycelium`, and `nix-shell` also worked.

**Comment by LeeSmet** [comment]
Thanks

**Pull request merged**